### PR TITLE
[Do Not Merge]move shared functionality b/w http and grpc to common folder

### PR DESCRIPTION
**Pull request by cbarik-infoblox**
To introduce an HTTP middleware which will authorise the incoming HTTP request by calling OPA sidecar, 
this PR aims to move the code that can be reused by the middleware to a common folder.
`Authorizer` interface will be custom implemented by `GRPC Authorizer` and `HTTP Authorizer`.
For the sake of being less intrusive, `GRPC Authorizer` is still named as `DefaultAuthorizer` as per the existing implementation.
All the grpc specific code will stay in `grpc_opa` folder, while the common code will move to `common` and its children folders.
image

### CODE REVIEW

File: common/authorizer/authorizer.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+type ClaimsVerifier func([]string, []string) (string, []error)
+
+// Authorizer interface is implemented for making arbitrary requests to Opa.
+type Authorizer interface {
+	// Evaluate is called with the grpc request's method passing the grpc request Context.
+	// If the handler is executed, the request will be sent to Opa. Opa's response
+	// will be unmarshaled using JSON into the provided response.
+	// Evaluate returns true if the request is authorized. The context
+	// will be passed to subsequent HTTP Handler.
+	Evaluate(ctx context.Context, fullMethod string, grpcReq interface{}, opaEvaluator OpaEvaluator) (bool, context.Context, error)
```
Comment (kanaksinghal) [inline common/authorizer/authorizer.go:19]: should we get rid of the "grpc" word references here?

**Review by rchowinfoblox — CHANGES_REQUESTED** [review]
While this code re-organization is good, atlas-authz-middleware is used by practically EVERY Infoblox microservice.  Moving the exported public API to different packages would break compatibility and force app-teams to expend effort to resolve if they upgrade their atlas-authz-middleware.

One suggestion is to put your re-organized code under a top-level atlas-authz-middleware/v2/ directory, and any new features will be under /v2/.  But this also means we have to maintain two separate versions of the middleware.

@drewwells @kumaya any comments on this?

**Comment by kumaya** [comment]
> While this code re-organization is good, atlas-authz-middleware is used by practically EVERY Infoblox microservice. Moving the exported public API to different packages would break compatibility and force app-teams to expend effort to resolve if they upgrade their atlas-authz-middleware.
> 
> One suggestion is to put your re-organized code under a top-level atlas-authz-middleware/v2/ directory, and any new features will be under /v2/. But this also means we have to maintain two separate versions of the middleware.
> 
> @drewwells @kumaya any comments on this?

Any service that is implementing `Authorizer` interface will need to upgrade the path if they choose to import the new release.
We can ~maintain~ create 2 versions, but would highly recommend to call out v1 (existing) as deprecated. So as to use any new functionality clients will be forced to upgrade the version; they can continue using the existing functionality without breaking things; and dev efforts will be saved maintaining 2 versions.

**Comment by cbarik-infoblox** [comment]
> While this code re-organization is good, atlas-authz-middleware is used by practically EVERY Infoblox microservice. Moving the exported public API to different packages would break compatibility and force app-teams to expend effort to resolve if they upgrade their atlas-authz-middleware.
> 
> One suggestion is to put your re-organized code under a top-level atlas-authz-middleware/v2/ directory, and any new features will be under /v2/. But this also means we have to maintain two separate versions of the middleware.
> 
> @drewwells @kumaya any comments on this?

@kumaya @rchowinfoblox  @drewwells raised a new PR https://github.com/infobloxopen/atlas-authz-middleware/pull/30/files

**Pull request closed without merging**